DOCS-10420 Microsoft Teams Example (#1363)
mulesoft/docs-connectors: microsoft-teams/1.0/modules/ROOT/pages/microsoft-teams-connector-examples.adoc

= Microsoft Teams Connector 1.0 Examples - Mule 4

> These examples shows you how to use Microsoft Teams Connector to perform the following actions:

Reviewer: ```suggestion
These examples show you how to use Microsoft Teams Connector to perform the following actions:
```

* <<create-new-team,Create a new team>>
* <<retrieve-team,Retrieve team properties and relationships>>
* <<retrieve-membership,Retrieve a list of team members>>
* <<retrieve-channel-list,Retrieve a list of team channels>>
* <<onboard-new-hire,Onboard a new hire to the team>>
* <<delete-channel,Delete a team channel>>
* <<remove-team-member,Remove a team member>>

== Prerequisites

To run these examples, you must have:

* A working instance of Microsoft Teams
* Credentials to access the Microsoft Teams instance
* Microsoft Teams Connector 1.0 or later
* Two users in the Microsoft Teams instance: one team owner and one user
* Anypoint Studio 7.x

== Configure OAuth Authorization

Because this connector uses OAuth to secure connections, you must specify OAuth field values, such as client credentials, keys, and OAuth-related URLs. To do this:

> * Create a configuration file for the OAuth Connection

Reviewer: ```suggestion
* Create a configuration file for the OAuth connection
```

* Configure the OAuth-related global elements
* Configure a global element for the properties file

=== Create a Configuration File for the OAuth Connection

> Create a configuration file for the OAuth connection:

Reviewer: Can you think of a way to tweak the steps intro to make it not exactly repeat the title?

. Create a file named `mule-app.properties` in the `/src/main/resources/` folder.

> . In the `mule-app.properties` file, create a set of properties for `Authorization Code Grant Type` like the ones that follows. Overlay the bracketed text (including the brackets) with the correct values for your OAuth configuration:

Reviewer: ```suggestion
. In the `mule-app.properties` file, create a set of properties for `Authorization Code Grant Type`, similar to the ones that follow, overlaying the bracketed text (including the brackets) with the correct values for your OAuth configuration:
```

Reviewer: I often substitute "similar to" for "like" for translation/globalization considerations.

+
----
config.oauth.authorization.code.consumerKey=<consumerKey>
config.oauth.authorization.code.consumerSecret=<consumerSecret>
config.oauth.authorization.code.authorizationUrl=<authorizationCode>
config.oauth.authorization.code.accessTokenUrl=<accessTokenUrl>
----
+

> . In the `mule-app.properties` file, create a set of properties for the `Client Credentials Grant Type` like the one that follows. Overlay the bracketed text (including the brackets) with the correct values for your OAuth configuration:

Reviewer: ```suggestion
. In the `mule-app.properties` file, create a set of properties for the `Client Credentials Grant Type`, similar to the one that follows, overlaying the bracketed text (including the brackets) with the correct values for your OAuth configuration:
```

+
----
config.oauth.client.credentials.clientId=<clientId>
config.oauth.client.credentials.clientSecret=<clientSecret>
config.oauth.client.credentials.tokenUrl=<tokenUrl>
----

> For more information about creating a properties file, see xref:mule-runtime::mule-app-properties-to-configure.adoc[Configuring Property Placeholders] for information.

Reviewer: ```suggestion
For more information about creating a properties file, see xref:mule-runtime::mule-app-properties-to-configure.adoc[Configuring Property Placeholders].
```

|===
. Click *OK*.
. Click the *Global Elements* tab and click *Create*.
. Select *Connector Configuration > Microsoft Teams Client Credentials Config* and click *OK*.
. Enter the following values:
+
[%header%autowidth.spread]
|===
| Field | Value
| Name | `Microsoft_Teams_Client_Credentials_Config`
| Client id |  `${config.oauth.client.credentials.clientId}`
| Client secret | `${config.oauth.client.credentials.clientSecret}`
| Token url | `${config.oauth.client.credentials.tokenUrl}`
| Scopes | Leave the default: `+https://graph.microsoft.com/.default+`
|===
. Click *OK*.

=== Configure a Global Element for the Properties File

Configure a global element for the `mule-app.properties` file so that Mule knows where to find it:

. Click the *Global Elements* tab and click *Create*.
. In the *Choose Global Type* dialog, select *Configuration properties* and click *OK*.
. In the *File* field, enter `mule.app.properties`.
. Click *OK*.

[[create-new-team]]
== Create a New Team

The following screenshot shows the Studio app flow for creating a new team:

.Use a flow like this one to create a new team.
image::ms-teams-create-team.png[Create a new team flow]

> Creating a new team involves configuring an HTTP *Listener* component, *Transform Message* component, *Create Team* operation, and second *Transform Message* component.

Reviewer: ```suggestion
Creating a new team involves configuring an HTTP *Listener* component, a *Transform Message* component, a *Create Team* operation, and a second *Transform Message* component.
```

To create the flow:

. Create a new Mule project in Studio.

> . In the Mule Palette view, search for *HTTP* and select the *Listener* operation:

Reviewer: ```suggestion
. In the Mule Palette view, search for *HTTP* and select the *Listener* operation.
```

. Drag the *Listener* operation onto the canvas.
. In the *Listener* configuration, click *+* next to the *Connector configuration* field to add a global element.
. Accept the defaults.
. In the HTTP properties window, set the *Path* field to `/createTeam`.

> === Add the first Transform Message Component

Reviewer: ```suggestion
=== Add the First Transform Message Component
```

The first *Transform Message* component creates a template for the input used to create the team:

> . In the Mule Palette view, search for *transform message*:

Reviewer: ```suggestion
. In the Mule Palette view, search for *transform message*.
```

. Drag the *Transform Message* component onto the canvas, to the right of the *Listener* component.

> . In the *Transform Message* configuration, overlay the brackets in the *Output* section with this XML:

Reviewer: Is there a reason to use "overlay" rather than replace? This is the first time that I have seen this phrasing in our docs is the reason I ask.

+
[source,dataweave,linenums]
----
{
	"template@odata.bind": "https://graph.microsoft.com/v1.0/teamsTemplates('standard')",
	description: message.attributes.queryParams.description,
	displayName: message.attributes.queryParams.displayName,
	"members":[
      {
        "@odata.type":"#microsoft.graph.aadUserConversationMember",
        "user@odata.bind":"https://graph.microsoft.com/v1.0/users('" ++ message.attributes.queryParams.user as String ++ "')",
         "roles":[
            "owner"
         ]
      }
}
----

=== Add the Create Team Operation

> The *Create team* operation creates the new team based on user input:

Reviewer: "the" or "a" team?

. Drag the *Create team* operation onto the canvas, to the right of the *Transform Message* component.

> . In the Create team configuration, click the  *Connector configuration* dropdown and select *Microsoft-Teams-Client-Config*.

Reviewer: Should "the Create team configuration" have some formatting/partial formatting?

Reviewer: dropdown > menu

. Select `Microsoft_Teams_Client_Credentials_Config` as the global element type and click *OK*.

> . Configure the following fields in the Create team properties window:

Reviewer: formatting in "the Create team properties window"?

+
[%header%autowidth.spread]
|===
|Field |Value
|Display Name |`Create team`
|Connector Configuration |`OAuth_Client_Credentials_Config`
|Message |`payload`
|===

=== Add the Second Transform Message Component

This *Transform Message* component converts the output of the *Create team* operation to JSON format.

> . In the Mule Palette view, search for *transform message*:

Reviewer: ```suggestion
. In the Mule *Palette* view, search for `transform message`:
```

. Drag the *Transform Message* component onto the canvas, to the right of the *Listener* component.
. Click the *Transform Message* component and set the output to `application/json`:
+
[source,dataweave,linenums]
----
%dw 2.0
output application/json
----

[[retrieve-team]]
== Retrieve the Team Properties and Relationships

Create a second flow to retrieve the properties and relationships for the new team. Use the *Get Team* operation in this flow:

.Use a flow like this one to retrieve the new team's properties and relationships.
image::ms-teams-get-team.png[Retrieve the team Flow]

[[retrieve-members]]
== Retrieve the Team Members

Create a third flow to retrieve information about the members of the new team. Use the *List team members* operation in this flow:

.Use a flow like this one to retrieve information about team members.
image::ms-teams-get-members.png[Retrieve the team members flow]

[[retrieve-channel-list]]
== Retrieve the Team Channels

> Create a fourth flow to retrieve a information about the channels used by the team. Use the *List channels* operation in this flow:

Reviewer: ```suggestion
Create a fourth flow to retrieve information about the channels used by the team. Use the *List channels* operation in this flow:
```

.Use a flow like this one to retrieve the channels used by the new team.
image::ms-teams-list-channels.png[Retrieve the team channels flow]

[[onboard-new-hire]]
== Onboard a New Hire to the Team

Create a fifth flow to onboard a new hire to the team.
Use the following operations in this flow:

* *Add team member* to add a new member to the team
* *Create channel* to create a new channel

> * *Add channel member* to add the new member to the newly-created channel

Reviewer: ```suggestion
* *Add channel member* to add the new member to the newly created channel
```

* *Create message* to create the welcome message

.Use a flow like this one to onboard a new user.
image::ms-teams-onboarding.png[Onboarding a new hire flow]

[[delete-channel]]
== Delete a Team Channel

Create a sixth flow to delete a channel. Use the *Delete channel* operation in this flow.

.Use a flow like this one to delete a channel.
image::ms-teams-delete-channel.png[Delete a channel]

[[remove-team-member]]
== Remove a Team Member

> Create a seventh flow to remove a member from a team. Use the *Remove team member* operation in this flow.

Reviewer: Does this numbering (th) of flows serve a purpose? Will the audience need to count?

.Use a flow like this one to remove a team member.
image::ms-teams-remove-member.png[Remove a team member]

== Run the App

To run the app:

. Right-click the project in Package Explorer and select *Run As > Mule Application*.
. After the app deploys, open a web browser.

> . Enter the following URL to start the OAuth dance: `+http://localhost:8081/authorize+`

Reviewer: ```suggestion
. Enter the following URL to start the OAuth dance: `+http://localhost:8081/authorize+`.
```

> . On the login screen, enter the login information used to access Microsoft Teams and click *Login*.

Reviewer: ```suggestion
. In the login screen, enter the login information used to access Microsoft Teams and click *Login*.
```

. Click *Allow*.
. Initiate a flow by entering the associated URL, as shown in the following table. If the URL has query parameters, ensure that you include the parameter values:
+
[%header%autowidth.spread]
|===
| Flow | URL | Notes
| Create a new team| `+http://localhost:8081/createTeam?displayName={teamDisplayName}&description={teamDescription}&user={teamOwnerUser}+` |
| Retrieve the team properties and relationships | `+http://localhost:8081/getTeam?team={createdTeamId}+` |
| Retrieve the team channels | `+http://localhost:8081/listChannels+` | Returns only the default channel because this example does not create channels.
| Onboard a new hire to the team| `+http://localhost:8081/newHireFlow?channelName={channelName}&team={createdTeamId}&userToOnboard={userToBeOnboarded}&channelOwner={channelOwner}+` a|

* For the `userToOnboard` query parameter, specify a user who is not the channel owner.

> * The JSON response contains the following welcome message: `"content": "Welcome to the team {channelName}"`

Reviewer: Why are these formatted as list items?

| Delete a team channel | `+http://localhost:8081/deleteChannel?team={teamId}&channel={channelId}+` |
| Remove a team member | `+http://localhost:8081/removeTeamMembers?team={teamId}&member={membershipId}+` |
|===

== XML for the Examples

> Paste this XML code into the *Configuration XML* tab in your project to experiment with the flows described in the previous sections. When you paste this code, click *Yes* on the Regenerate 'doc:id' Values dialog.

Reviewer: ```suggestion
Paste this XML code into the *Configuration XML* tab in your project to experiment with the flows. When you paste this code, click *Yes* in the *Regenerate 'doc:id' Values* dialog.
```

* <<retrieve-membership,Retrieve a list of team members>>
* <<retrieve-channel-list,Retrieve a list of team channels>>
* <<onboard-new-hire,Onboard a new hire to the team>>
* <<delete-channel,Delete a team channel>>
* <<remove-team-member,Remove a team member>>

== Prerequisites

To run these examples, you must have:

* A working instance of Microsoft Teams
* Credentials to access the Microsoft Teams instance
* Microsoft Teams Connector 1.0 or later
* Two users in the Microsoft Teams instance: one team owner and one user
* Anypoint Studio 7.x

== Configure OAuth Authorization

Because this connector uses OAuth to secure connections, you must specify OAuth field values, such as client credentials, keys, and OAuth-related URLs. To do this:

* Create a configuration file for the OAuth connection
* Configure the OAuth-related global elements
* Configure a global element for the properties file

=== Create a Configuration File for the OAuth Connection

Create a configuration file for the OAuth connection that includes properties for the `Authorization Code` and `Client Credentials` grant types:

. Create a file named `mule-app.properties` in the `/src/main/resources/` folder.

> . In the `mule-app.properties` file, create a set of properties for `Authorization Code` grant type, similar to the ones that follow, replacing the bracketed text (including the brackets) with the correct values for your OAuth configuration:

Reviewer 2: ```suggestion
. In the `mule-app.properties` file, create a set of properties for the `Authorization Code` grant type, similar to the ones that follow, replacing the bracketed text (including the brackets) with the correct values for your OAuth configuration:
```

. In the Mule Palette view, search for *transform message*.
. Drag the *Transform Message* component onto the canvas, to the right of the *Listener* component.
. In the *Transform Message* configuration, overlay the brackets in the *Output* section with this XML:
+
[source,dataweave,linenums]
----
{
	"template@odata.bind": "https://graph.microsoft.com/v1.0/teamsTemplates('standard')",
	description: message.attributes.queryParams.description,
	displayName: message.attributes.queryParams.displayName,
	"members":[
      {
        "@odata.type":"#microsoft.graph.aadUserConversationMember",
        "user@odata.bind":"https://graph.microsoft.com/v1.0/users('" ++ message.attributes.queryParams.user as String ++ "')",
         "roles":[
            "owner"
         ]
      }
}
----

=== Add the Create Team Operation

The *Create team* operation creates a new team based on user input:

. Drag the *Create team* operation onto the canvas, to the right of the *Transform Message* component.
. In the Create team configuration, click the  *Connector configuration* dropdown and select *Microsoft-Teams-Client-Config*.

> . Select `Microsoft_Teams_Client_Credentials_Config` as the global element type and click *OK*.

Reviewer 2: If you are selecting that in the UI, shouldn't it be bold?
